### PR TITLE
Add AWS cloud provider option for IAM role

### LINKED ISSUE #59526: Run Cloud Controller Manager with AWS IAM role from different account
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind feature
/sig aws

**Anything else we need to know?**:
Enable running the Cloud Controller Manager using a different AWS account than the worker instances.

The AWS cloud provider for Kubernetes assumes that all of the master and worker nodes run in the same AWS account. I would like to be able to run the master nodes in a separate AWS account to keep separate the running of the cluster from what runs on the cluster. By default, the cloud provider uses the EC2 instance role. I would like a way to provide an AWS IAM role to use instead.

For example, if I created my master instance in AWS account A, and my worker instance in account B, I want to be able to provide an IAM role from account B that Controller Manager, running on an instance in account A, would assume. Any resources created (I.E. ELB) would be created using the assume role in account B.

I would like to propose adding a RoleArn property to the AWS cloud config file. If present, the cloud provider would use an AssumeRoleProvider instead of the EC2RoleProvider when initializing the AWS SDK client such that all requests to AWS are made with the assumed role. AssumeRoleProvider automatically renews the assumed role when it expires. 

**Also considered:**
I could use the environment variables to pass temporary credentials after manually assuming the target role (AWS_ACCESS_KEY_ID, etc). As far as I understand, the credentials returned from assuming a role expire after a certain amount of time so this would not work for a long-running process. The Cloud Controller Manager process would need to be restarted in order to re-assume the role after it has expired.

Note this request was previously shared here for initial comments here:
https://docs.google.com/document/d/1x4DI30Q3YI8Sr3x4XNZ3dcjISL1rj2KbVnM1ld5S4KA/edit?usp=sharing

### PR DESCRIPTION
**Pull request by brycecarman**
**What this PR does / why we need it**:
Adds the option to provide an IAM role ARN in the AWS cloud provider config file that should be assumed when communicating with the AWS APIs. 
For example, this allows running Controller Manager in a account separate from the worker nodes, but still allows all resources created to interact with the workers. ELBs created would be in the same account as the worker nodes for instance. 

**Which issue(s) this PR fixes** *(optional, in fixes #<issue number>(, fixes #<issue_number>, ...) format, will close the issue(s) when PR gets merged)*:
Fixes #59526

**Special notes for your reviewer**:
None

**Release note**:

```release-note
Add AWS cloud provider option to use an assumed IAM role 
```

### CODE REVIEW

File: pkg/cloudprovider/providers/aws/aws.go [hunk 1]
```diff
@@ -526,6 +528,9 @@ type CloudConfig struct {
 		// RouteTableID enables using a specific RouteTable
 		RouteTableID string
 
+		// The IAM role to assume when interacting with AWS
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws.go]: Nit: I believe go comments are supposed to follow "RoleArn is the IAM role..."

>  Doc comments work best as complete sentences, which allow a wide variety of automated presentations. The first sentence should be a one-sentence summary that starts with the name being declared.

https://golang.org/doc/effective_go.html

I'm not that strict on it though :-)

File: pkg/cloudprovider/providers/aws/aws.go [hunk 2]
```diff
@@ -927,22 +932,43 @@ func (s *awsSdkEC2) DescribeVpcs(request *ec2.DescribeVpcsInput) (*ec2.DescribeV
...
+		var provider credentials.Provider
+		if cfg.Global.RoleArn == "" {
+			provider = &ec2rolecreds.EC2RoleProvider{
+				Client: ec2metadata.New(sess),
+			}
+		} else {
+			glog.Infof("Using AWS assumed role %v", cfg.Global.RoleArn)
+			provider = &stscreds.AssumeRoleProvider{
+				Client:  sts.New(sess),
+				RoleARN: cfg.Global.RoleArn,
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws.go]: On RoleArn vs RoleARN - Go tends to prefer `RoleARN`.  https://github.com/golang/go/wiki/CodeReviewComments#initialisms

But this is actually bound to a config file.  It looks like you can override the mapping with a `gcfg` tag, but we haven't done that yet.  Maybe just be sure that you want the config option in the file to be `RoleArn` (i.e. is that consistent with usage elsewhere in AWS configuration files, outside of kubernetes) and then I think consistency trumps golang naming conventions.

File: pkg/cloudprovider/providers/aws/aws_test.go [hunk 3]
```diff
@@ -389,7 +396,7 @@ func TestGetRegion(t *testing.T) {
 
 func TestFindVPCID(t *testing.T) {
 	awsServices := newMockedFakeAWSServices(TestClusterId)
-	c, err := newAWSCloud(strings.NewReader("[global]"), awsServices)
+	c, err := newAWSCloud(CloudConfig{}, awsServices)
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws_test.go]: This makes me happy :-)

**Comment by justinsb** [comment]
/ok-to-test

/lgtm

Maybe double-check if you want the config file option to be RoleArn or RoleARN, but I'll go with what you recommend.

File: pkg/cloudprovider/providers/aws/aws.go (hunk 2 above)
Comment (brycecarman) [reply pkg/cloudprovider/providers/aws/aws.go]: Agreed. It is RoleARN in AWS SDK so there is no reason to deviate from that.

**Comment by justinsb** [comment]
/lgtm

Thanks @brycecarman !

**Pull request merged**